modules/ding_gatewayf/ding_gatewayf.module

```diff
@@ -0,0 +1,458 @@
+
+  // Redirect the user to gatewayf for logout.
+  $items[DING_GATEWAYF_LOGOUT_URL] = array(
+    'page callback' => 'ding_gatewayf_logout',
+    'access callback' => 'user_is_logged_in',
+    'type' => MENU_CALLBACK,
+  );
+
+  // Gatewayf callback which is limited to anonymous users to prevent extra
+  // login and redirects.
+  $items[DING_GATEWAYF_CALLBACK_URL] = array(
+    'page callback' => 'ding_gatewayf_callback',
+    'access callback' => TRUE,
+    'type' => MENU_CALLBACK,
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_theme().
+ */
+function ding_gatewayf_theme($existing, $type, $theme, $path) {
+  return array(
+    'ding_gatewayf_login' => array(
+      'template' => 'ding-gatewayf-login',
+      'variables' => array(
+        'link' => NULL,
+      ),
+      'path' => $path . '/templates',
+    ),
+  );
+}
+
+/**
+ * Implements hook_block_info().
+ */
+function ding_gatewayf_block_info() {
+  return array(
+    'login' => array(
+      'info' => t('Ding GateWAYF login link'),
+    ),
+  );
+}
+
+/**
+ * Implements hook_block_view().
+ */
+function ding_gatewayf_block_view() {
+  $block = new stdClass();
+
+  // Only display the link when users are not logged in.
+  if (user_is_anonymous()) {
+    $img = array(
+      '#theme' => 'image',
+      '#path' => drupal_get_path('module', 'ding_gatewayf') . '/images/nemid_logo.png',
+      '#alt' => t('NemID logo'),
+      '#title' => t('NemID logo'),
+    );
+
+    $block->content = array(
+      '#theme' => 'ding_gatewayf_login',
+      '#link' => array(
+        '#theme' => 'link',
+        '#text' => t('Login with !logo', array('!logo' => drupal_render($img))),
+        '#path' => DING_GATEWAYF_LOGIN_URL,
+        '#options' => array(
+          // The attributes array is required or drupal_render will return WSOD.
+          'attributes' => array(),
+          'query' => array(
+            drupal_get_destination(),
+          ),
+          'html' => TRUE,
+        ),
+      ),
+    );
+  }
+
+  return $block;
+}
+
+/**
+ * Implements hook_user_logout().
+ */
+function ding_gatewayf_user_logout() {
+  // Check if the user used gatewayf to login, there is no need to redirect the
+  // user if gatewayf was not used.
+  if (_ding_gatewayf_is_logged_in_with_gatewayf()) {
+    ding_gatewayf_logout();
+  }
+}
+
+
+/**
+ * Redirect the user to login page at the gatewayf service.
+ */
+function ding_gatewayf_login() {
+  $config = _ding_gatewayf_load_configuration();
+
+  // Check if destination is set and if it is encode it into the callback URL so
+  // the user can be redirect to correct page after login at gatewayf.
+  $callback = $config['service']['callback'];
+  if (!empty($_REQUEST['destination'])) {
+    $callback = urlencode($callback . '?destination=' . $_REQUEST['destination']);
+  }
+
+  // Redirect the user by setting the location header.
+  header('Location:' . $config['service']['endpoint'] . '?returnUrl=' . $callback . '&idp=' . $config['service']['idp']);
+  exit;
```

Comment:
Yes, drupal_goto() will not work and I do what to stop the execution of Drupal. I see no need to invoke the hook_exists and use resources on that here.